packages/components/src/internal/components/user/UserLink.tsx

```diff
@@ -1,30 +1,107 @@
 
 import { User } from '../base/models/User';
-import { AppURL } from '../../url/AppURL';
+import { userCanReadUserDetails } from '../../app/utils';
 
-interface Props {
-    allUsers: List<IUser>;
-    currentUser: User;
-    userId: string;
+import { caseInsensitive } from '../../util/utils';
+
+import { selectRowsUserProps, UserDetailsPanel } from './UserDetailsPanel';
+import {useServerContext} from "../base/ServerContext";
+
+interface UserLinkProps {
+    userDisplayValue?: string;
+    userId: number;
+    unknown?: boolean;
 }
 
-export class UserLink extends React.PureComponent<Props> {
-    render() {
-        const { allUsers, currentUser, userId } = this.props;
-        if (!allUsers || !userId) return null;
-
-        let targetUser: IUser = null;
-        if (allUsers) {
-            targetUser = allUsers.find(user => user.userId === parseInt(userId));
-        }
-
-        if (targetUser) {
-            const link = AppURL.create('q', 'core', 'siteusers', userId).toHref();
-            return currentUser.isAdmin ? <a href={link}>{targetUser.displayName}</a> : targetUser.displayName;
-        }
-        return null;
+export const UserLink: FC<UserLinkProps> = props => {
+    const { userId, userDisplayValue, unknown } = props;
+    const [showDetails, setShowDetails] = useState<boolean>(false);
+    const [targetUserDisplayValue, setTargetUserDisplayValue] = useState<string>();
+    const { user, container } = useServerContext();
+    const isSelf = userId === user.id;
+
+    useEffect(() => {
+        (async () => {
+            try {
+                if (!!userId && userDisplayValue === undefined) {
+                    const targetUser = await selectRowsUserProps(userId);
+                    setTargetUserDisplayValue(caseInsensitive(targetUser, 'DisplayName'));
+                } else {
+                    setTargetUserDisplayValue(userDisplayValue);
+                }
+            } catch (error) {
+                console.error(error);
+            }
+        })();
+    }, [userDisplayValue, userId]);
+
+    const toggleDetailsModal = useCallback(() => {
+        setShowDetails(current => !current);
+    }, []);
+
+    if (targetUserDisplayValue && !userId) return <span>{targetUserDisplayValue}</span>;
+
+    if (unknown) {
+        return (
+            <span className="gray-text" title="User may have been deleted from the system.">
+                &lt;unknown user&gt;
+            </span>
+        );
+    }
+
+    if (!userId) return null;
+
+    if (!isSelf && (!userCanReadUserDetails(user) || !targetUserDisplayValue)) {
+        if (targetUserDisplayValue) return <span>{targetUserDisplayValue}</span>;
+        return (
+            <span className="gray-text" title="User may have been deleted from the system.">
+                &lt;{userId}&gt;
+            </span>
+        );
     }
+
+    return (
+        <>
+            <a onClick={toggleDetailsModal} style={{ cursor: 'pointer' }}>
```

Comment:
I think there's a `clickable` class for this style.

Reply:
hurray! thank you.